### ISSUE TITLE
angular vscode extension based on stryker-karma-runner

### ISSUE THREAD

**Issue by Raagh** [stryker-mutator/stryker#1305]
Hey guys, so I am trying to build a visual studio code extension that allows me to easily run angular tests and see outcomes, I found your code online and it might be what I need but I am having trouble to make it work. Basically I need to load angular the angular cli, start karma and run it using a custom reporter. this way I can get the tests and their results and show them in a pretty ui on vscode.

do you have something in mind that I need to change just to manually run your code? I edited the code a little to remove the logger and the stryker dependencies. I am using the same karma.conf.js that in the project and I am passing this as fake runner options export default class RunnerOptions {
  public port: any = 9876;
  public strykerOptions: any = {
    karma: {
      projectType: "angular-cli",
      ngConfig: {},
    },
  };
  public ngOptions: any;
}

to KarmaTestRunner, unfortunately after the return of the cli object in the angularStarter it just goes to error with exitCode 1 and no extra information.

here is a copy of the code I am using in case you wanna try it yourself(please follow the readme) https://github.com/Raagh/angular-test-explorer

Again sorry to bother you guys and thanks

**Comment by Raagh** [comment]
Hi @nicojs , thanks for anwsering. Yes I am already taking a look at this. My problem is I am not quite sure how I should be running the angular tests programatically. thats my biggest issue the rest I can deal with it. In your code I see that you re invoking the angular cli and then from what i see using karma.runner directly. Could you explain to me how you guys are starting the angular testing workflow directly from the code?

**Comment by nicojs** [comment]
Sure.

We start angular from this file: https://github.com/stryker-mutator/stryker/blob/master/packages/stryker-karma-runner/src/starters/angularStarter.ts#L15

We're using the kind-of-private api of the angular cli. I tried to make it public here: https://github.com/angular/angular-cli/pull/12499. Unfortunately they didn't want it open.

So basically just use `const cli = require('@angular/cli')`  and then use `cli({
    cliArgs,
    inputStream: process.stdin,
    outputStream: process.stdout
})`. The way to configure the test run is by creating a custom `karma.conf.js` file and use that. That way you can override whatever configuration you want. You can still choose to load the user's custom karma config from there. You can look at our file for a reference: https://github.com/stryker-mutator/stryker/blob/master/packages/stryker-karma-runner/src/starters/stryker-karma.conf.ts

The way to pass data to the custom karma.conf.js file is with static variables, that's what the `setGlobals` function is for: https://github.com/stryker-mutator/stryker/blob/master/packages/stryker-karma-runner/src/starters/stryker-karma.conf.ts#L123.

**Comment by Raagh** [comment]
Hey thanks for the great explanation, it was really helpful. I was definitely on the right track. I decided to start from scratch so I can try to make it work. First I am trying to run programatically the same configuration as doing ng test on an actual angular project. So I managed to run the cli command, and it returns exitcode 0.  but I noticed a few things: . If I only run the cli and then try to do karma.runner.run() it will say that there is no server listening on that port. if I do cli -> karma.server start -> karma.runner.run() there is no error but I am still not seeing any of the tests of the angular project being run. in your project which path do you take? you just do cli and then karma run? or the 3 complete steps? with the default angular cli configuration should I be seeing the tests run in a chrome browser right?

**Comment by nicojs** [comment]
Yes, we start karma and then run multiple times with `karma.runner.run`. However, this static method needs the port of the karma server. This is the method we use to run karma: https://github.com/stryker-mutator/stryker/blob/3a5e65f25c0962078eaccfaa250f16f9dc447d83/packages/stryker-karma-runner/src/KarmaTestRunner.ts#L133-L140

It doesn't give you the test result, only resolves when the run started successfully. You then need to listen to the karma process for events. We use our custom reporter for that: https://github.com/stryker-mutator/stryker/blob/master/packages/stryker-karma-runner/src/StrykerReporter.ts

**Comment by Raagh** [comment]
Hey Nico, sorry for keep asking you things. Feel free to start ignoring me at any time.  So I been playing this for the last few days, unfortunately I didn't made success with the cli approach. I tried running the cli as you mentioned, looking at your code and the cli itself but unfortunately nothing happens. It exists with return code 0 but still the ng test command is just not running. I think this is probably related to the fact that I am trying to run the cli from a directory outside the angular project(even though I am using the cli node_module installed in the angular project).

I took a different approach. I am creating a commandline with basepath  = Angularproject root. and running the command ng test passing --karma-config: my karma config

this runs the command correctly of course, after that I could run karma with karma.run multiple times.
my new problem  is that I dont understand how can I pass data to my karma config

I saw your stryker-karma.conf.ts but I dont understand how that works, cause if you import that file and use setGlobals to pass some values, but then you tell the cli to use the compiled js file what happens to the values you passed using the .ts import? I am probably missing something big there.
